File: src/Jering.DynamicForms/DynamicInput.cs
using System.Collections.Generic;

namespace Jering.DynamicForms
{
    /// <summary>
    /// 
    /// </summary>
    public class DynamicInput
    {
        /// <summary>
        /// 
        /// </summary>
        public string Name { get; set; }
        /// <summary>
        /// 
        /// </summary>
        public string DisplayName { get; set; }
        /// <summary>
        /// 
        /// </summary>
        public string InitialValue { get; set; }
        /// <summary>
        /// email, password etc
        /// </summary>
        public string Type { get; set; }
        /// <summary>
        /// input, select etc
        /// </summary>
        public string TagName { get; set; }
        /// <summary>
        /// 
        /// </summary>
        public List<DynamicInputValidatorData> ValidatorData { get; set; }
        /// <summary>
        /// 
        /// </summary>
        public int Order { get; set; }
        /// <summary>
        /// 
        /// </summary>
        public string Placeholder { get; set; }
        /// <summary>
        /// 
        /// </summary>
        public bool RenderLabel { get; set; }
    }
}

File: src/jering-vectorartkit-webclient/src/vak-app/shared/dynamic-forms/dynamic-input/dynamic-input.ts
import { DynamicInputValidatorData } from './dynamic-input-validator-data';
import { DynamicInputValidator } from './dynamic-input-validators';
import { DynamicForm } from '../dynamic-form/dynamic-form';

/**
 * @Description
 * 
 */
export class DynamicInput<T>{
    initialValue: T;
    name: string;
    displayName: string;
    order: number;
    type: string;
    tagName: string;
    placeholder: string;
    renderLabel: boolean;
    validatorData: DynamicInputValidatorData[];

    errors: string[];
    parent: DynamicForm;
    validators: DynamicInputValidator[];
    value: string;
    dirty: boolean;
    blurred: boolean;

    constructor(options: {
        initialValue?: T,
        name?: string,
        displayName?: string,
        order?: number,
        type?: string,
        tagName?: string,
        placeholder?: string,
        renderLabel?: boolean,
        validatorData?: DynamicInputValidatorData[]
    } = {}) {
        this.initialValue = options.initialValue;
        this.name = options.name || '';
        this.displayName = options.displayName || '';
        this.order = options.order === undefined ? 1 : options.order;
        this.type = options.type || '';
        this.tagName = options.tagName || '';
        this.placeholder = options.placeholder || '';
        this.renderLabel = options.renderLabel || false;
        this.validatorData = options.validatorData || [];
    }

    validate(): void {
        if (this.dirty && this.blurred) {
            this.errors = [];
            for (let validator of this.validators) {
                let error = validator(this);
                if (error != null) {
                    this.errors.push(error);
                }
            }
        }
    }
}
